libkbfs/md_journal.go

```diff
@@ -0,0 +1,507 @@
+			brmd.Revision, brmd.PrevRoot)
+
+		if i > 0 {
+			log.Debug("Changing prev root of rev=%s to %s",
+				brmd.Revision, prevID)
+			brmd.PrevRoot = prevID
+		}
+
+		newID, err := j.putMD(brmd, currentUID, currentVerifyingKey)
+		if err != nil {
+			return err
+		}
+
+		o, err := revisionToOrdinal(brmd.Revision)
+		if err != nil {
+			return err
+		}
+
+		err = j.j.j.writeJournalEntry(o, newID)
+		if err != nil {
+			return err
+		}
+
+		prevID = newID
+
+		log.Debug("Changing ID for rev=%s from %s to %s",
+			brmd.Revision, id, newID)
+	}
+
+	return err
+}
+
+func (j mdJournal) pushEarliestToServer(
+	ctx context.Context, log logger.Logger, signer cryptoSigner,
+	mdserver MDServer) (MdID, *BareRootMetadata, error) {
+	earliestID, err := j.j.getEarliest()
+	if err != nil {
+		return MdID{}, nil, err
+	}
+	if earliestID == (MdID{}) {
+		return MdID{}, nil, nil
+	}
+
+	rmd, err := j.getMD(earliestID)
+	if err != nil {
+		return MdID{}, nil, err
+	}
+
+	log.Debug("Flushing MD put id=%s, rev=%s", earliestID, rmd.Revision)
+
+	var rmds RootMetadataSigned
+	rmds.MD = *rmd
+	err = signMD(ctx, j.codec, signer, &rmds)
+	if err != nil {
+		return MdID{}, nil, err
+	}
+	err = mdserver.Put(ctx, &rmds)
+	if err != nil {
+		// Still return the ID and RMD so that they can be
+		// consulted.
+		return earliestID, rmd, err
+	}
+
+	return earliestID, rmd, nil
+}
+
+// All functions below are public functions.
+
+func (j mdJournal) length() (uint64, error) {
+	return j.j.length()
+}
+
+func (j mdJournal) getHead(currentUID keybase1.UID) (
+	ImmutableBareRootMetadata, error) {
+	return j.checkGetParams(currentUID)
+}
+
+func (j mdJournal) getRange(
+	currentUID keybase1.UID, start, stop MetadataRevision) (
+	[]ImmutableBareRootMetadata, error) {
+	_, err := j.checkGetParams(currentUID)
+	if err != nil {
+		return nil, err
+	}
+
+	realStart, mdIDs, err := j.j.getRange(start, stop)
+	if err != nil {
+		return nil, err
+	}
+	var rmds []ImmutableBareRootMetadata
+	for i, mdID := range mdIDs {
+		expectedRevision := realStart + MetadataRevision(i)
+		rmd, err := j.getMD(mdID)
+		if err != nil {
+			return nil, err
+		}
+		if expectedRevision != rmd.Revision {
+			panic(fmt.Errorf("expected revision %v, got %v",
+				expectedRevision, rmd.Revision))
+		}
+		irmd := MakeImmutableBareRootMetadata(rmd, mdID)
+		rmds = append(rmds, irmd)
+	}
+
+	return rmds, nil
+}
+
+// MDJournalConflictError is an error that is returned when a put
+// detects a rewritten journal.
+type MDJournalConflictError struct{}
+
+func (e MDJournalConflictError) Error() string {
+	return "MD journal conflict error"
+}
+
+func (j mdJournal) put(
+	ctx context.Context, signer cryptoSigner, ekg encryptionKeyGetter,
+	rmd *RootMetadata, currentUID keybase1.UID,
+	currentVerifyingKey VerifyingKey) (MdID, error) {
+	head, err := j.getHeadHelper()
+	if err != nil {
+		return MdID{}, err
+	}
+
+	mStatus := rmd.MergedStatus()
+	bid := rmd.BID
+
+	if (mStatus == Unmerged) && (bid == NullBranchID) &&
```

Comment:
Maybe a comment here or above the function that it's ok to pass in an unmerged id without the branch ID set, since it gets fixed up here as long as the unmerged flag is set.